File: common/src/main/java/net/impleri/playerskills/api/TeamMode.java
package net.impleri.playerskills.api;

enum TeamModeType {
    OFF,
    SHARED,
    SPLIT_EVENLY,
    PYRAMID,
    PROPORTIONAL,
    LIMITED,
}

public class TeamMode {
    public static TeamMode off() {
        return new TeamMode(TeamModeType.OFF);
    }

    public static TeamMode shared() {
        return new TeamMode(TeamModeType.SHARED);
    }

    public static TeamMode splitEvenly() {
        return new TeamMode(TeamModeType.SPLIT_EVENLY);
    }

    public static TeamMode pyramid() {
        return new TeamMode(TeamModeType.PYRAMID);
    }

    public static TeamMode proportional(Double percentage) {
        return new TeamMode(TeamModeType.PROPORTIONAL, percentage);
    }

    public static TeamMode limited(Double amount) {
        return new TeamMode(TeamModeType.LIMITED, amount);
    }

    private final TeamModeType type;
    public final Double rate;

    private TeamMode(TeamModeType type, Double rate) {
        this.type = type;
        this.rate = rate;
    }

    private TeamMode(TeamModeType type) {
        this(type, null);
    }

    public boolean isOff() {
        return type == TeamModeType.OFF;
    }

    public boolean isShared() {
        return type == TeamModeType.SHARED;
    }

    public boolean isSplitEvenly() {
        return type == TeamModeType.SPLIT_EVENLY;
    }

    public boolean isPyramid() {
        return type == TeamModeType.PYRAMID;
    }

    public boolean isProportional() {
        return type == TeamModeType.PROPORTIONAL;
    }

    public boolean isLimited() {
        return type == TeamModeType.LIMITED;
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/api/TeamMode.kt
package net.impleri.playerskills.api

class TeamMode private constructor(private val type: TeamModeType, val rate: Double? = null) {
  val isOff: Boolean
    get() = type == TeamModeType.OFF
  val isShared: Boolean
    get() = type == TeamModeType.SHARED
  val isSplitEvenly: Boolean
    get() = type == TeamModeType.SPLIT_EVENLY
  val isPyramid: Boolean
    get() = type == TeamModeType.PYRAMID
  val isProportional: Boolean
    get() = type == TeamModeType.PROPORTIONAL
  val isLimited: Boolean
    get() = type == TeamModeType.LIMITED

  override fun toString(): String {
    return type.name
  }

  companion object {
    fun off(): TeamMode {
      return TeamMode(TeamModeType.OFF)
    }

    fun shared(): TeamMode {
      return TeamMode(TeamModeType.SHARED)
    }

    fun splitEvenly(): TeamMode {
      return TeamMode(TeamModeType.SPLIT_EVENLY)
    }

    fun pyramid(): TeamMode {
      return TeamMode(TeamModeType.PYRAMID)
    }

    fun proportional(percentage: Double): TeamMode {
      return TeamMode(TeamModeType.PROPORTIONAL, percentage)
    }

    fun limited(amount: Double): TeamMode {
      return TeamMode(TeamModeType.LIMITED, amount)
    }
  }
}
